jni/src/Foreign/JNI.cpphs

```diff
@@ -392,31 +401,27 @@ jvm = unsafePerformIO $ alloca $ \pjvm -> alloca $ \pnum_jvms -> do
 --
 -- Yields @Nothing@ if the calling thread is not attached to the JVM.
 getJNIEnv :: IO (Maybe (Ptr JNIEnv))
-getJNIEnv = do
-    env0 <- [CU.exp| JNIEnv* { jniEnv } |]
-    if nullPtr /= env0 then
-      return $ Just env0
-    else do
-      rc <- [CU.exp| jint {
-          (*$(JavaVM* jvm))->GetEnv($(JavaVM* jvm), (void**)&jniEnv, JNI_VERSION_1_6)
-        }|]
-      when (rc == [CU.pure| jint { JNI_EVERSION } |]) $
-        fail "GetEnv: Unsupported version."
-      if rc == [CU.pure| jint { JNI_EDETACHED } |] then
-        return Nothing
-      else do
-        when (rc /= jni_ok) $
-          throwIO $ JNIError "GetEnv" rc
-        env <- [CU.exp| JNIEnv* { jniEnv } |]
-        return $ Just env
+getJNIEnv = [CU.exp| JNIEnv* { jniEnv } |] >>= \case
+  env | env /= nullPtr -> return $ Just env
+  _ -> do
+    throwIfNotOK_
+      [CU.exp| jint {
+        (*$(JavaVM* jvm))->GetEnv($(JavaVM* jvm), (void**)&jniEnv, JNI_VERSION_1_6)
+      }|]
+    env <- [CU.exp| JNIEnv* { jniEnv } |]
+    return $ Just env
 
 -- | Run an action against the appropriate 'JNIEnv'.
 --
 -- Each OS thread has its own 'JNIEnv', which this function gives access to.
 --
 -- TODO check whether this call is only safe from a (bound) thread.
 withJNIEnv :: (Ptr JNIEnv -> IO a) -> IO a
-withJNIEnv f = maybe (throwIO ThreadNotAttached) f =<< getJNIEnv
+withJNIEnv f = do
+    env <- [CU.exp| JNIEnv* { jniEnv } |]
+    if env == nullPtr
+    then throwIO ThreadNotAttached
+    else f env
```

Comment:
I don't think we can dispense with the call to getJNIEnv, when Java calls into Haskell, the thread is already attached, yet, `jniEnv` hasn't been set.

Reply:
Hm, but that amounts to saying the invariant as currently stated can't hold?